FIXME enable 64bit bitmapset and update visimap

yaowangm opened the pull request:
It's split from https://github.com/greenplum-db/gpdb/pull/14529. It includes the portion about enabling 64bit bms and other necessary changes related to visimap.

### Summary

Postgresql supports bitmapset uses 32 bit or 64 hit implementation. GP uses bitmapset as data store of AppendOnlyVisimapEntry. However, the implementation of visimap is based on 32bit only. The result is that bitmapset on GP is limited to 32bit, even on a 64bit platform. The PR is to enable 64bit bitmapset on 64bit platform.

### Solution

The trick is, visimap uses 32bit for not only in-memory data but also on-disk data. Code to encode/decode on-disk data (`Bitmap_Compress`/`BitmapDecompress_Decompress`) accept an uint32 array (`visiMapEntry->bitmap->words`) and encode/decode data in uint32 unit. However, `visiMapEntry->bitmap->words` is an uint64 array if bms is in 64bit.

Finally I adapted a kind of hybrid solution: 

1. Enable 64bit bitmapset (on 64bit platform), which will benefit all relevant operations.
2. visimap uses 64bit bitmapset in memory.
3. Keep encode/decode functions unchanged, i.e. they will encode/decode data in 32bit as before.

The most important thing is the join between 2 and 3. The interfaces to encode/decode on-disk data (`Bitmap_Compress`/`BitmapDecompress_Decompress`) accepts an uint32 array and a word count as input. So we can forcely convert a uint64 array to uint32 array, with doubled word count, as input. The approach can keep the encoding/decoding part unchanged, and take full advantage of 64bit bms otherwhere.

Another reason to keep the encoding/decoding part unchanged is, it works on both 32bit/64bit platform, and I don't want to implement a separated 64bit encoding/decoding part.

### Bitmap index

Bitmap index also uses bitmap data structure, but it implemented an independent bit stream who uses 64bit words always:

```
/*
 * The size in bits of a hybrid run-length(HRL) word.
 * If you change this, you should change the type for
 * a HRL word as well.
 */
#define BM_HRL_WORD_SIZE        64

/* the type for a HRL word */
typedef uint64   
```
So it should not be impacted. 

### Others

In the FIXME comments Heikki also mentioned "using bitmapword for TID bitmaps, just like upastream, and eliminate tbm_bitmapword". It seems not very necessary because that tbm_bitmapword is 64bit currently, and we don't have much motivation to support 32bit feature.

### Test

The relevant tests includes 3 scenarios:

1. on 64bit env: The code didn't change any existing user interface, so the scenario has been covered by existing tests, including bitmap related UTs on src/backend/utils/misc/test.
2. on 32bit env: I forcedly compiled 32bit bms by hacking code, and issued all the tests.
3. 32bit env -> 64bit env: it means creating ondisk data on 32bit env and reading the data on 64bit env (e.g. upgrading GPDB6 to GPDB7). Since existing tests don't cover pgupgrade, I had to manually issue the tests.

yaowangm commented:
Perf test:

As talked, I referred the tests from the discussion: https://www.postgresql.org/message-id/flat/CAKJS1f9EGBd2h-VkXvb%3D51tf%2BX46zMX5T8h-KYgXEV_u2zmLUw%40mail.gmail.com

There are two tests: the first one got only 0.89% average increase, which is due to most cost was spent on IO operation. The second one got much better increase 8.23% by another query. Unfortunately, the script for test 2 is not provided.

I reproduced the test 1 based on the script provided. It's the result:
```
32bit mode:
58.537280
59.181717
59.264590
59.185271
59.386761

64bit mode:
58.834336
58.653824
59.076401
58.936742
58.290691
```
The result is similar and the difference (0.6%) is even a little bit smaller. Additional cost for distributed arch on GPDB should explain the smaller difference. Another query heavily using bms is supposed to show bigger difference like test 2. But I am not sure whether it's worth more work since the test has shown small but steady improvement on GPDB similar to postgresql if we exclude the most cost for IO operations.

Please let me know your comments, thanks.

yaowangm reviewed (commented):
More update abour perf test:

All previous tests I did spent most cost on IO so did not get significant improvement. Finally I hacked the code of bitmap_compression_test.c and added a function `test__BitmapCompression__BmsPerf()`. It accesses bms in memory directly by `bms_add_member()` and `bms_is_member()` and avoid any IO cost. This time I got ideal result.

Code:
```
static void 
test__BitmapCompression__BmsPerf(void **state) 
{
    Bitmapset *bms = NULL;

    printf("bms_times = %d\n", bms_times);
    for (int i = 0;i < bms_times;i++)
    {   
        bms = bms_add_member(bms, i);
    }

    for (int i = 0;i < bms_times;i++)
    {
        if(!bms_is_member(i, bms))
            printf("error at %d\n", i);
    }
    bms_free(bms);
    printf("good\n");
}
```
Test result (seconds, bms_times = 200000000 for each round):

32bit mode:
```
1.774s
1.768s
1.774s
1.770s
1.771s
Sum: 8.857s
```
64bit mode:
```
1.453s
1.454s
1.452s
1.455s
1.451s
Sum: 7.265s
```
The avg improvement is **17.97%**. I didn't test other bms functions and believe we have very solid proof for the performance improvement now.

soumyadeep2007 commented on src/backend/access/appendonly/appendonly_visimap_entry.c:
```diff
@@ -264,23 +298,84 @@ AppendOnlyVisimapEntry_GetHiddenTupleCount(
+		if (BITS_PER_BITMAPWORD == 64)
+		{
+			/*
+			 * On 64bit env, if there is only one word in bms, and the last half
+			 * of the word is empty, it means there is only one 32bit word
+			 * actually.
```
Please reword (and format):
```
/*
 * On 64bit env, if there is only one 64 bit word in memory, and the 32 higher order bits
 * of that word are all zero, it implies that there is only one 32 bit word.
 * We can always assume that the 32 higher order bits for a 64 bit bitmap word is zeroed
 * out - this is ensured by routines such as bms_add_member() and
 * AppendOnlyVisiMapEnty_ReadData().
 */
```

soumyadeep2007 commented on src/backend/access/appendonly/appendonly_visimap_entry.c:
```diff
@@ -264,23 +298,84 @@ AppendOnlyVisimapEntry_GetHiddenTupleCount(
+	Assert(bmsWordCount <= APPENDONLY_VISIMAP_MAX_BITMAP_WORD_COUNT);
 
 	Assert(visiMapEntry->data);
 	Assert(APPENDONLY_VISIMAP_DATA_BUFFER_SIZE >= bitmapSize);
+	/*
+	 * On production environment without assertion, we need to terminate
+	 * current backend if we hit the error.
+	 */
+	if (bitmapSize > APPENDONLY_VISIMAP_DATA_BUFFER_SIZE)
+	{
+		elog(FATAL,
+			 "incorrect bitmapSize: "
+			 "APPENDONLY_VISIMAP_DATA_BUFFER_SIZE = %lu, "
+			 "bitmapSize = %d, "
+			 "visiMapEntry->bitmap->nwords = %d",
+			 APPENDONLY_VISIMAP_DATA_BUFFER_SIZE,
+			 bitmapSize,
+			 visiMapEntry->bitmap->nwords);
+	}
```
If we have the FATAL, we don't need the Assert.

yaowangm replied:
Update

soumyadeep2007 commented on src/backend/access/appendonly/appendonly_visimap_entry.c:
```diff
@@ -499,6 +597,20 @@ AppendOnlyVisimapEntry_GetMinimalSizeToCover(int64 offset)
 	minSize |= minSize >> 16;
 	minSize++;
+
+	Assert(minSize <= APPENDONLY_VISIMAP_MAX_BITMAP_WORD_COUNT);
+	/*
+	 * On production environment without assertion, we need to terminate
+	 * current backend if we hit the error.
+	 */
+	if (minSize > APPENDONLY_VISIMAP_MAX_BITMAP_WORD_COUNT)
+	{
+		elog(FATAL,
+			 "incorrect minSize: offset=%ld, minSize=%u",
+			 offset,
+			 minSize);
+	}
```
If we have the FATAL, we don't need the Assert.

soumyadeep2007 commented on src/backend/utils/misc/bitmap_compression.c:
```diff
@@ -105,74 +115,11 @@ BitmapDecompress_Decompress(BitmapDecompressState *state,
-				rleRepeatCount--;
-			}
-		}
-		if (rleRepeatCount > 0)
-		{
-			elog(ERROR, "illegal RLE state after bitmap decompression: "
-					"block count %d, compression type %d, rle repeat count %u",
-					state->blockCount, state->compressionType, rleRepeatCount);
-		}
-
-		if (failed)
-		{
-			elog(ERROR, "bitstream read error seen during decompression: "
-					"block count %d, compression type %d",
-					state->blockCount, state->compressionType);
-		}
+		Bitmap_Compress_DefaultDecompress(state, bitmap);
```
I thought we would be doing the refactoring (if at all we need to) in a different PR.

soumyadeep2007 commented on src/backend/access/appendonly/appendonly_visimap_entry.c:
```diff
@@ -156,21 +159,52 @@ AppendOnlyVisiMapEnty_ReadData(AppendOnlyVisimapEntry *visiMapEntry, size_t data
 	visiMapEntry->bitmap = NULL;
-	newWordCount =
+	onDiskBlockCount =
 		BitmapDecompress_GetBlockCount(&decompressState);
-	if (newWordCount > 0)
+	/* The on-disk bitmap representation always uses 32-bit block size
+	 * (for backward compatibility). Depending on the environment, we
+	 * may be using either * 64-bit words or 32-bit words for the
+	 * in-memory representation.
+	 * So, if (in-memory) bitmapset uses 64 bit words, we can use half
+	 * of the on-disk bitmap block count.
+	 */
+	if (BITS_PER_BITMAPWORD == 64)
+	{
+		/*
+		 * Number of on-disk blocks is always 0, 1 or even.
+		 * See resizing logic in AppendOnlyVisimapEntry_HideTuple()
+		 */
+		if (onDiskBlockCount == 1)
+		{
+			bmsWordCount = 1;
+		}
+		else
+		{
+			Assert(onDiskBlockCount % 2 == 0);
+			bmsWordCount = onDiskBlockCount / 2;
+		}
+	}
+	else
+	{
+		Assert(BITS_PER_BITMAPWORD == 32);
+		bmsWordCount = onDiskBlockCount;
+	}
+	Assert(bmsWordCount <= APPENDONLY_VISIMAP_MAX_BITMAP_WORD_COUNT);
+	Assert(bmsWordCount >= 0);
+
+	if (onDiskBlockCount > 0)
```
I discussed the PR with @ashwinstar and he had the following suggestion:
1. Extract this logic into a new function: `BitmapDecompress_GetBlockCounts`.
2. Use the 2 block counts obtained for the palloc0 and call to `BitmapDecompress_Decompress` below.
3. Unit test BitmapDecompress_GetBlockCounts with cmockery.

soumyadeep2007 commented on src/backend/access/appendonly/appendonly_visimap_entry.c:
```diff
@@ -264,23 +298,86 @@ AppendOnlyVisimapEntry_GetHiddenTupleCount(
+		/*
+		 * On 64bit env, there is a conflict: in-memory bms is in 64bit word,
+		 * but on-disk block is in 32bit word to keep consistency. We need to
+		 * provide 32bit block count to Bitmap_Compress() after kind of
+		 * conversion.
+		 */
+		if (BITS_PER_BITMAPWORD == 64)
+		{
+			/*
+			 * On 64bit env, if there is only one 64 bit word in memory, and the
+			 * 32 higher order bits of that word are all zero, it implies that
+			 * there is only one 32 bit word. We can always assume that the 32
+			 * higher order bits for a 64 bit bitmap word is zeroed out - this
+			 * is ensured by routines such as bms_add_member() and 
+			 * AppendOnlyVisiMapEnty_ReadData().
+			 */
+			if (bmsWordCount == 1
+				&& (visiMapEntry->bitmap->words[0] >> 32) == 0)
+			{
+				blockCount = 1;
+			}
+			else
+			{
+				/*
+				 * blockCount required by Bitmap_Compress() is always in
+				 * uint32-words. So, if bitmapset uses 64 bit words, double
+				 * the value of bmsWordCount.
+				 */
+				blockCount = visiMapEntry->bitmap->nwords * 2;
+			}
+		}
+		else
+		{
+			/*
+			 * On 32bit env, blockCount is always equal to bmsWordCount.
+			 */
+			blockCount = visiMapEntry->bitmap->nwords;
+		}
+	}
+	bitmapSize = sizeof(uint32) * blockCount;
```
1. Add a similar function to extract the logic to get the block count.
2. Use that for the call to Bitmap_Compress below.
3. Unit test the new function added with cmockery.

soumyadeep2007 commented on src/include/nodes/bitmapset.h:
```diff
@@ -31,24 +31,15 @@ struct List;
- * appendoptimized tables during upgrade (not viable).  The goal of this fixme
- * is to explore if the performance benefit of larger bitmapwords can still be
- * availed by distinguishing on-disk bitmap usage from strictly in-memory
- * bitmap usage.  E.g. define a new type bitmapword32 and use it in
- * appendoptimized code.  On a related note, tbm_bitmapword is Greenplum
- * specific 64-bit wide type used for TID bitmaps.  Can we start using
- * bitmapword for TID bitmaps, just like upastream, and eliminate
- * tbm_bitmapword?
```
Discussed w/ @ashwinstar :
Let's leave part of this FIXME here. I was wrong to suggest that we can't replace tbm_bitmapword with bitmapword. We can. There is no upgrade impact. We should tackle this in a following PR.

yaowangm commented:
@soumyadeep2007 I have addressed the comments. Please take another look, thx!
Something in `bitmap_compression_test.c`: since there is no context about visimap and related bitmapset, I have to create a bitmapset with the bitmap data by kind of hacking (just like the existing code did to create bitmap data). A side effect is, I have to move the definition of `BITMAPSET_SIZE` from `bitmapset.c `to `bitmapset.h`, which is different from upstream. Please let me know if you have better idea. Thanks!

soumyadeep2007 commented on src/backend/utils/misc/bitmap_compression.c:
```diff
@@ -357,3 +364,106 @@ Bitmap_Compress(
+			 * is ensured by routines such as bms_add_member() and
+			 * AppendOnlyVisiMapEnty_ReadData().
+			 */
+			if (*bmsWordCount == 1
+				&& (bitmap->words[0] >> 32) == 0)
+			{
+				*onDiskBlockCount = 1;
+			}
```
Also do we have a unit test covering this case?

soumyadeep2007 reviewed (commented):
> Something in bitmap_compression_test.c: since there is no context about visimap and related bitmapset, I have to create a bitmapset with the bitmap data by kind of hacking (just like the existing code did to create bitmap data). A side effect is, I have to move the definition of BITMAPSET_SIZE from bitmapset.c to bitmapset.h, which is different from upstream. Please let me know if you have better idea. Thanks!

All of this is perfectly fine. Thank you for adding tests and making the code better!

This is looking better. Mostly cosmetic changes suggested, one important question regarding the logic and one test improvement suggestion.

soumyadeep2007 commented on src/backend/utils/misc/test/bitmap_compression_test.c:
```diff
@@ -17,6 +17,27 @@ test__BitmapCompression__ZeroBitmap(void **state)
+
 	int r = Bitmap_Compress(
 		BITMAP_COMPRESSION_TYPE_DEFAULT, 
 		bitmap, 4,
```
> Writing separate functions doesn't mean much here. Tests of _CalculateBlockCounts() are to test the output of the two functions for various of scenarios: one block, more blocks, with/without actual compression, etc. So we need to add the test code of the two functions into various existing test cases rather then creating separate functions. Correct?

Philosophically, the only code you have changed in this PR is that you have added a new mechanism to calculate block counts before passing them to Compress/Decompress routines. You haven't changed the Compress/Decompress routines, and they don't depend on the `_CalculateBlockCounts` routines. So it would have been sufficient to test ONLY the block count calculation code you have written.

The tests you have added are more in the integration style, which isn't bad.

I'm happy as long we are testing `_CalculateBlockCounts` with all variations of input, which is what is happening right now in the PR, indirectly.

soumyadeep2007 reviewed (approved):
LGTM!

yaowangm commented on src/backend/utils/misc/test/bitmap_compression_test.c:
```diff
@@ -281,18 +453,37 @@ test__BitmapCompression_ShortDecompress(void **state)
+	Bitmapset *bms;
+
+	/* fake a bitmapset with the bitmap data */
+	bms = (Bitmapset *) palloc0(BITMAPSET_SIZE(expectedBmwWordCount));
```
Thanks. I updated all callings of `palloc0()` except one in `test__BitmapCompression__IllegalCompressionType()`. For the case, we actually copied only the first 32bit block. If we don't init the memory of bms, we will miss the conidition `bitmap->words[0] >> 32) == 0` and generate incorrect result. pretty subtle :-)

soumyadeep2007 replied:
Fair point. Thanks for keeping an eye on that!
